### PR TITLE
Add the compile build tool

### PR DESCRIPTION
**Pull request by ianthehat**
This uses the same style as the asm tool, and removes the need for filter_exec

### CODE REVIEW

File: go/tools/builders/compile.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	}
+	gotool := args[0]
+	args = args[1:]
+	sources := []string{}
+	goargs := []string{}
+	bctx := build.Default
+	bctx.CgoEnabled = true
+	for i, s := range args {
+		if s == "--" {
+			if i < len(args) {
```
Comment (jayconrod) [inline go/tools/builders/compile.go]: I think you can get rid of this condition. It should always be true.

File: go/tools/builders/compile.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+				goargs = args[i+1:]
+			}
+			break
+		}
+		sources = append(sources, s)
+	}
+	// filter the soure list to the ones that apply
```
Comment (jayconrod) [inline go/tools/builders/compile.go]: Typo in "soure". But consider rewording to "Apply build constraints to the source list".

File: go/tools/builders/compile.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+func run(args []string) error {
+	// process the args
+	if len(args) < 2 {
+		return fmt.Errorf("Usage: compile gotool [sources] -- <extra options>")
+	}
+	gotool := args[0]
+	args = args[1:]
+	sources := []string{}
+	goargs := []string{}
```
Comment (jayconrod) [inline go/tools/builders/compile.go]: Consider renaming to `goopts`. Keep the `goargs` as a separate variable below when you build the full argument list with `tool`, `compile` and sources.

File: go/tools/builders/compile.go (hunk 2 above)
Comment (ianthehat) [inline go/tools/builders/compile.go]: done

File: go/tools/builders/compile.go (hunk 1 above)
Comment (ianthehat) [inline go/tools/builders/compile.go]: It happens if the -- is the last entry in the list

File: go/tools/builders/compile.go (hunk 3 above)
Comment (ianthehat) [inline go/tools/builders/compile.go]: done

**Pull request merged**